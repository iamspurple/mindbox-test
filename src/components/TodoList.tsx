import type { ITodoItem } from "../types";

const TodoList = ({
  list,
  toggleCompleted,
}: {
  list: ITodoItem[];
  toggleCompleted: (id: number) => void;
}) => {
  return (
    <ul>
      {list.map((item) => (
        <li key={item.id}>
          <input
            type="checkbox"
            checked={item.completed}
            onChange={() => toggleCompleted(item.id)}
          />
          {item.title}
        </li>
      ))}
    </ul>
  );
};

export default TodoList;
