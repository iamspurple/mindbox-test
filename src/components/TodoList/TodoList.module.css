.list {
  list-style: none;
  font-family: "ZenKaku";
  font-weight: 300;
  text-align: left;
  padding: 0;
  box-shadow: 0px 3px 10px 3px rgba(34, 60, 80, 0.08);
}

.item {
  background-color: rgba(255, 255, 255);
  padding: 0.6rem;
  border-bottom: 1px solid rgba(230, 223, 223, 0.5);
  display: flex;
  align-items: center;
  gap: 1rem;
}

.completed {
  text-decoration: line-through;
  color: grey;
}

.checkbox {
  width: 1rem;
  height: 1rem;
  appearance: none;
  position: relative;

  border-radius: 50%;
  border: 1px solid #b5b5b5;
  transition: 500ms;
}

.checkbox:checked::after {
  content: "";
  position: absolute;
  bottom: 1px;
  left: 3px;
  width: 1rem;
  aspect-ratio: 1;
  background: url("../../../public/icons/checkmark.svg") center/contain
    no-repeat;
}
