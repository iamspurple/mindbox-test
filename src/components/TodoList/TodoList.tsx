import type { ITodoItem } from "../../types";
import styles from "./TodoList.module.css";

const TodoList = ({
  list,
  toggleCompleted,
}: {
  list: ITodoItem[];
  toggleCompleted: (id: string) => void;
}) => {
  return (
    <ul className={styles.list}>
      {list.map((item) => (
        <li key={item.id} className={styles.item}>
          <input
            className={styles.checkbox}
            type="checkbox"
            checked={item.completed}
            onChange={() => toggleCompleted(item.id)}
          />
          <span className={item.completed ? styles.completed : ""}>
            {item.title}
          </span>
        </li>
      ))}
    </ul>
  );
};

export default TodoList;
