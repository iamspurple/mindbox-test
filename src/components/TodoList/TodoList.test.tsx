import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';
import type { ITodoItem } from '../../types';

jest.mock('./TodoList.module.css', () => ({
    list: 'list',
    item: 'item',
    checkbox: 'checkbox',
    completed: 'completed',
}));

describe('TodoList', () => {
    const mockToggleCompleted = jest.fn();
    const sampleTodos: ITodoItem[] = [
        { id: '1', title: 'First todo', completed: false },
        { id: '2', title: 'Second todo', completed: true },
    ];

    beforeEach(() => {
        mockToggleCompleted.mockClear();
    });

    it('renders list of todos', () => {
        render(<TodoList list={sampleTodos} toggleCompleted={mockToggleCompleted} />);

        expect(screen.getByText('First todo')).toBeInTheDocument();
        expect(screen.getByText('Second todo')).toBeInTheDocument();
        expect(screen.getAllByRole('listitem').length).toBe(2);
    });

    it('applies completed class for completed items', () => {
        render(<TodoList list={sampleTodos} toggleCompleted={mockToggleCompleted} />);

        const completedItem = screen.getByText('Second todo');
        const activeItem = screen.getByText('First todo');

        expect(completedItem).toHaveClass('completed');
        expect(activeItem).not.toHaveClass('completed');
    });

    it('renders checkboxes with correct checked state', () => {
        render(<TodoList list={sampleTodos} toggleCompleted={mockToggleCompleted} />);

        const checkboxes = screen.getAllByRole('checkbox');
        expect(checkboxes[0]).not.toBeChecked();
        expect(checkboxes[1]).toBeChecked();
    });

    it('calls toggleCompleted with correct id when checkbox is clicked', () => {
        render(<TodoList list={sampleTodos} toggleCompleted={mockToggleCompleted} />);

        const firstCheckbox = screen.getAllByRole('checkbox')[0];
        fireEvent.click(firstCheckbox);

        expect(mockToggleCompleted).toHaveBeenCalledTimes(1);
        expect(mockToggleCompleted).toHaveBeenCalledWith('1');
    });

    it('renders empty state when list is empty', () => {
        render(<TodoList list={[]} toggleCompleted={mockToggleCompleted} />);

        expect(screen.queryByRole('listitem')).not.toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(
            <TodoList list={sampleTodos} toggleCompleted={mockToggleCompleted} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});