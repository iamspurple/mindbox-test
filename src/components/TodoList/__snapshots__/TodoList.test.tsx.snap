// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoList matches snapshot 1`] = `
<DocumentFragment>
  <ul
    class="list"
  >
    <li
      class="item"
    >
      <input
        class="checkbox"
        type="checkbox"
      />
      <span
        class=""
      >
        First todo
      </span>
    </li>
    <li
      class="item"
    >
      <input
        checked=""
        class="checkbox"
        type="checkbox"
      />
      <span
        class="completed"
      >
        Second todo
      </span>
    </li>
  </ul>
</DocumentFragment>
`;
